### PR TITLE
feat: Add renderHtmlDataCells to SimpleTable.ts

### PR DESCRIPTION
**Pull request by jbrundage**
Signed-off-by: Jaman Brundage <jbrundage372@gmail.com>

## Checklist:
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit message includes a "fixes" reference if appropriate.
  - [x] The commit is signed.
- [x] The change has been fully tested:
  - [x] I have viewed all related gallery items
  - [ ] I have viewed all related dermatology items
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised new issues to address them separately

### CODE REVIEW

File: packages/html/src/SimpleTable.ts [hunk 1]
```diff
@@ -1,6 +1,9 @@
 import { HTMLWidget } from "@hpcc-js/common";
 import { select as d3Select } from "d3-selection";
 
+function entitiesEncode(str) {
```
Comment (GordonSmith) [inline packages/html/src/SimpleTable.ts]: Missing typings

Comment (GordonSmith) [reply packages/html/src/SimpleTable.ts]: Duplicate code with:  https://github.com/hpcc-systems/Visualization/blob/b6817413f1d7a7aef2ee9115c43d05b54f81fdb9/packages/eclwatch/src/WUResultStore.ts#L4-L20 (common up and move to util.

File: packages/html/src/SimpleTable.ts [hunk 2]
```diff
@@ -33,10 +37,11 @@ export class SimpleTable extends HTMLWidget {
             .append("th")
             .attr("class", (n, i) => `th-${i}`)
             .merge(theadTrSelection)
-            .text(_d => (_d).toString())
+        [insertMethodName](_d => (_d).toString())
```
Comment (GordonSmith) [inline packages/html/src/SimpleTable.ts]: This will bypass typescript compile time checking

File: packages/html/src/SimpleTable.ts [hunk 3]
```diff
@@ -47,7 +52,8 @@ export class SimpleTable extends HTMLWidget {
                     .append("td")
                     .attr("class", (n, i) => `col-${i}`)
                     .merge(tdSelection as any)
-                    .text(_d => (_d).toString())
+                    .attr("title", _d => cellDataOnHover ? entitiesEncode((_d).toString()) : "")
+                [insertMethodName](_d => (_d).toString())
```
Comment (GordonSmith) [inline packages/html/src/SimpleTable.ts]: This will bypass typescript compile time checking

File: packages/html/src/SimpleTable.ts [hunk 4]
```diff
@@ -58,7 +64,13 @@ export class SimpleTable extends HTMLWidget {
...
 
 export interface SimpleTable {
+    renderHtmlDataCells(): boolean;
+    renderHtmlDataCells(_: boolean): this;
+    cellDataOnHover(): boolean;
+    cellDataOnHover(_: boolean): this;
     autoWidth(): boolean;
     autoWidth(_: boolean): this;
 }
+SimpleTable.prototype.publish("cellDataOnHover", false, "boolean", "When hovering over a TD display the full cell content");
```
Comment (GordonSmith) [inline packages/html/src/SimpleTable.ts]: I wouldn't make this optional.

**Comment by jbrundage** [comment]
@GordonSmith this is ready for review

File: packages/html/src/SimpleTable.ts [hunk 5]
```diff
@@ -25,16 +26,21 @@ export class SimpleTable extends HTMLWidget {
...
+        const theadTrUpdate = theadTrSelection.enter()
             .append("th")
             .attr("class", (n, i) => `th-${i}`)
             .merge(theadTrSelection)
-            .text(_d => (_d).toString())
             ;
+        if (this.renderHtmlDataCells()) {
+            theadTrUpdate.html(d => (d).toString());
+        } else {
+            theadTrUpdate.text(d => (d).toString());
```
Comment (GordonSmith) [inline packages/html/src/SimpleTable.ts]: Why is `(d)` in the parens?

File: packages/html/src/SimpleTable.ts [hunk 6]
```diff
@@ -58,7 +69,10 @@ export class SimpleTable extends HTMLWidget {
 SimpleTable.prototype._class += " html_SimpleTable";
 
 export interface SimpleTable {
+    renderHtmlDataCells(): boolean;
+    renderHtmlDataCells(_: boolean): this;
```
Comment (GordonSmith) [inline packages/html/src/SimpleTable.ts]: Does this match the equivalent name in Table and dgrid.Table?

File: packages/html/src/SimpleTable.ts (hunk 5 above)
Comment (jbrundage) [reply packages/html/src/SimpleTable.ts]: I'm not sure. It was that way when I started working on it and I left it that way. It should be harmless to remove it.

File: packages/html/src/SimpleTable.ts (hunk 6 above)
Comment (jbrundage) [reply packages/html/src/SimpleTable.ts]: It matches `other_Table`, but I don't think `dgrid_Table` has that feature yet.

Comment (GordonSmith) [reply packages/html/src/SimpleTable.ts]: Thats fine.

File: packages/html/src/SimpleTable.ts [hunk 7]
```diff
@@ -25,16 +26,21 @@ export class SimpleTable extends HTMLWidget {
...
         const theadTrSelection = this._theadRow.selectAll("th").data(this.columns());
-        theadTrSelection.enter()
+        const theadTrUpdate = theadTrSelection.enter()
             .append("th")
             .attr("class", (n, i) => `th-${i}`)
             .merge(theadTrSelection)
-            .text(_d => (_d).toString())
             ;
+        if (this.renderHtmlDataCells()) {
+            theadTrUpdate.html(d => d.toString());
```
Comment (GordonSmith) [inline packages/html/src/SimpleTable.ts]: Why the `toString`?

File: packages/html/src/SimpleTable.ts [hunk 8]
```diff
@@ -25,16 +26,21 @@ export class SimpleTable extends HTMLWidget {
...
+        } else {
+            theadTrUpdate.text(d => d.toString());
```
Comment (GordonSmith) [inline packages/html/src/SimpleTable.ts]: Why the `toString`?

File: packages/html/src/SimpleTable.ts [hunk 9]
```diff
@@ -43,12 +49,17 @@ export class SimpleTable extends HTMLWidget {
             .each(function (this, d) {
                 const tr = d3Select(this);
                 const tdSelection = tr.selectAll("td").data(d);
-                tdSelection.enter()
+                const tdUpdate = tdSelection.enter()
                     .append("td")
                     .attr("class", (n, i) => `col-${i}`)
                     .merge(tdSelection as any)
-                    .text(_d => (_d).toString())
+                    .attr("title", _d => safeEncode((_d).toString()))
```
Comment (GordonSmith) [inline packages/html/src/SimpleTable.ts]: Why the parenthesis around `_d`?
Why the `toString`?

File: packages/html/src/SimpleTable.ts [hunk 10]
```diff
@@ -43,12 +49,17 @@ export class SimpleTable extends HTMLWidget {
...
                     ;
+                if (context.renderHtmlDataCells()) {
+                    tdUpdate.html(_d => (_d).toString());
```
Comment (GordonSmith) [inline packages/html/src/SimpleTable.ts]: Why the parenthesis around _d?
Why the toString?

File: packages/html/src/SimpleTable.ts [hunk 11]
```diff
@@ -43,12 +49,17 @@ export class SimpleTable extends HTMLWidget {
...
+                } else {
+                    tdUpdate.text(_d => (_d).toString());
```
Comment (GordonSmith) [inline packages/html/src/SimpleTable.ts]: Why the parenthesis around _d?
Why the toString?

File: packages/util/src/string.ts [hunk 12]
```diff
@@ -19,3 +19,21 @@ export function endsWith(origString: string, searchString: string, position?: nu
     const lastIndex = subjectString.lastIndexOf(searchString, position);
     return lastIndex !== -1 && lastIndex === position;
 }
+
+export function entitiesEncode(str: string): string {
+    return String(str).replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
+}
```
Comment (GordonSmith) [inline packages/util/src/string.ts]: Not sure this should be exported?

File: packages/util/src/string.ts [hunk 13]
```diff
@@ -19,3 +19,21 @@ export function endsWith(origString: string, searchString: string, position?: nu
...
+        case "[object Boolean]":
+        case "[object Number]":
+            return item as undefined | boolean | number;
+        case "[object String]":
+            return entitiesEncode(item as string);
+        default:
+            console.log("Unknown cell type:  " + Object.prototype.toString.call(item));
+    }
+    return item;
+}
```
Comment (GordonSmith) [inline packages/util/src/string.ts]: Did you remove the other instance of this code?

**Pull request closed without merging**